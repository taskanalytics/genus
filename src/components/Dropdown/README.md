```react
state: {toggle: true}
---
<div>
  
  <button onClick={() => setState({toggle: !state.toggle})}>
    Toggle dropdown
  </button>

  <Dropdown
    toggle={state.toggle}
    close={() => setState({toggle: false})}
    actions={[
      {
        name: 'Option 1',
        type: '',
        action: () => console.log('Clicked Option 1'),
      },
      {
        name: 'Option 2',
        type: '',
        action: () => console.log('Clicked Option 2'),
      },
      {
        name: 'Option 3',
        type: '',
        action: () => console.log('Clicked Option 3'),
      },
      {
        name: 'Option 4',
        type: '',
        action: () => console.log('Clicked Option 4'),
      },
      {
        type: 'separator',
      },
      {
        name: 'Heading',
        type: 'heading',
      },
      {
        name: 'Option 5',
        type: '',
        action: () => console.log('Clicked Option 5'),
      },
      {
        name: 'Option 6',
        type: 'destructive',
        action: () => console.log('Clicked Option 6'),
      }
    ]} />
  
</div>
```
