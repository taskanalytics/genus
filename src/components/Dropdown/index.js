/** @jsx jsx */
import React, { Component } from 'react'
import { findDOMNode } from 'react-dom'
import T from 'prop-types'
import { jsx } from '@emotion/core'

import {
  StyledMobileClose,
  StyledHeading,
  StyledSeparator,
  StyledDropdown,
  StyledItem,
  StyledWrapper,
} from './styled'

class Dropdown extends Component {
  static defaultProps = {
    right: false,
    renderTrigger: props => (
      <button type='button' {...props}>
        Toggle
      </button>
    ),
  }

  state = {
    open: false,
  }

  closeMenu = () => {
    this.setState({ open: false }, () => {
      document.removeEventListener('click', this.onClickOutside)
    })
  }

  onClickOutside = event => {
    if (this.dd && !this.dd.contains(event.target)) {
      this.closeMenu()
    }
  }

  toggle = event => {
    event.preventDefault()
    const open = !this.state.open
    this.setState({ open }, () => {
      if (open) {
        document.addEventListener('click', this.onClickOutside)
      } else {
        document.removeEventListener('click', this.onClickOutside)
      }
    })
  }

  componentWillUnmount () {
    document.removeEventListener('click', this.onClickOutside)
  }

  renderActions = actions => {
    const passedProps = { close: this.closeMenu }
    if (typeof actions === 'function') {
      return actions({
        ...passedProps,
        Item: StyledItem,
      })
    }

    return actions.map((action, key) => {
      const props = {
        ...(action.props || {}),
        key: `action-${key}`,
        destructive: action.type === 'destructive',
      }
      if (action.render) {
        action.type = 'component'
      }
      switch (action.type) {
        case 'separator':
          return <StyledSeparator {...props} />
        case 'heading':
          return <StyledHeading {...props}>{action.name}</StyledHeading>
        case 'component':
          return (
            <StyledItem {...props}>
              <action.render />
            </StyledItem>
          )
        default:
          return (
            <StyledItem
              {...props}
              onClick={e => {
                e.stopPropagation()
                action.action(e, passedProps)
                this.closeMenu()
              }}
            >
              {action.name}
            </StyledItem>
          )
      }
    })
  }

  render () {
    const {
      right,
      actions,
      renderTrigger,
      dialogStyles = {},
      ...props
    } = this.props
    const { open } = this.state

    return (
      <StyledWrapper {...props} data-genus='Dropdown'>
        {renderTrigger({
          onClick: this.toggle,
        })}
        <StyledDropdown
          ref={ref => {
            this.dd = findDOMNode(ref)
          }}
          open={open}
          right={right}
        >
          <StyledMobileClose
            onClick={e => {
              e.stopPropagation()
              this.closeMenu()
            }}
          >
            X
          </StyledMobileClose>
          {this.renderActions(actions)}
        </StyledDropdown>
      </StyledWrapper>
    )
  }
}

Dropdown.propTypes = {
  actions: T.oneOfType([T.array, T.func]).isRequired,
  right: T.bool,
  renderTrigger: T.func,
  dialogStyles: T.object,
}

export default Dropdown
