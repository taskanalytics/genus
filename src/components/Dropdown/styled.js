import styled from '@emotion/styled'
import css from '@styled-system/css'
import { keyframes } from '@emotion/core'
import { rgba } from 'polished'
import { space, color, fontSize, compose } from 'styled-system'

import { Box } from '../Grid'
import Button from '../Button'

export const StyledWrapper = styled(Box)(
  css({
    position: ['initial', 'relative'],
    display: 'inline-block',
  }),
)

const fadeFromTop = keyframes({
  '0%': { opacity: 0, top: '40%' },
  '20%': { opacity: 0.2, top: '50%' },
  '40%': { opacity: 0.4, top: '70%' },
  '50%': { opacity: 0.8, top: '90%' },
  '100%': { opacity: 1, top: null },
})

const fadeFromBottom = keyframes({
  '0%': { bottom: '-90%' },
  '20%': { bottom: '-70%' },
  '40%': { bottom: '-50%' },
  '50%': { bottom: '-20%' },
  '100%': { bottom: null },
})

export const StyledDropdown = styled(Box)(
  ({ theme, open, right, direction = 'bottom' }) => {
    let styles = {}
    if (direction === 'top') {
      styles = {
        mb: [0, 1],
        bottom: [0, '100%'],
        right: [0, right ? 0 : 'unset'],
        left: [0, !right ? 0 : 'unset'],
        animation: `${fadeFromBottom} 100ms linear`,
      }
    } else {
      styles = {
        mt: [0, 1],
        bottom: [0, 'unset'],
        right: [0, right ? 0 : 'unset'],
        left: [0, !right ? 0 : 'unset'],
        animation: [
          `${fadeFromBottom} 100ms linear`,
          `${fadeFromTop} 100ms linear`,
        ],
      }
    }
    return css({
      position: ['fixed', 'absolute'],
      minWidth: [1, 160],
      bg: 'white',
      color: 'text',
      borderRadius: [0, 2],
      borderWidth: 0,
      padding: 1,
      boxShadow: [`0px 20px 20px 15px ${theme.colors.muted}`, 'neutral.small'],
      zIndex: 50,
      display: open ? 'flex' : 'none',
      flexDirection: 'column',
      justifyContent: 'start',
      '&:focus': {
        outline: 'none',
      },
      maxHeight: ['60%', 'unset'],
      overflow: 'auto',
      ...styles,
    })
  },
  color,
)

export const StyledMobileClose = styled(Button)(
  css({
    display: ['block', 'none'],
    justifySelf: 'flex-start',
    alignSelf: 'flex-end',
    fontSize: 3,
    border: 'none',
    p: 0,
    m: 3,
    color: 'primary',
    '&:hover': {
      boxShadow: 'none',
      border: 'none',
    },
  }),
)

export const StyledItem = styled('button')(
  ({ theme, destructive }) =>
    css({
      display: 'block',
      width: '100%',
      textAlign: 'left',
      backgroundColor: 'transparent',
      fontFamily: '"Brandon Text",sans-serif',
      fontSize: 1,
      fontWeight: 400,
      border: 'none',
      py: 1,
      px: 2,
      borderRadius: 1,
      cursor: 'pointer',
      color: destructive && theme.colors.trouble,
      '&:hover,&:focus,&:active': {
        backgroundColor: rgba(theme.colors.primary, 0.1),
      },
    }),
  compose(
    space,
    fontSize,
    color,
  ),
)

export const StyledHeading = styled(Box)(
  css({
    fontWeight: 500,
    py: 1,
    px: 2,
  }),
)

export const StyledSeparator = styled('hr')(
  css({
    bg: 'muted',
    border: 0,
    m: 1,
    height: 1,
  }),
)
