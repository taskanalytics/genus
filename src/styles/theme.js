import { rgba } from 'polished'
import { colors } from './colors'

const theme = {
  breakpoints: [
    '32em', '48em', '64em',
  ],
  space: [
    0, 6, 12, 18, 24, 48,
  ],
  radius: 4,
  fontSizes: [
    12, 16, 18, 24, 36, 72
  ],
  colors: {
    black: 'rgb(17, 17, 17)',
    blue: '#07c',
    text: 'rgb(45, 54, 68)',
    white: '#fff',
    muted: 'rgba(45, 54, 68, 0.6)',
    primary: colors.primary,
    primaryShaded: colors.primaryShaded,
    good: rgba(140, 215, 150, 1),
    goodDark: '#3BBE77',
    headsUp: rgba(255, 212, 0, 1),
    headsUpDark: '#FFA707',
    trouble: rgba(255, 90, 20, 1),
    troubleDark: '#FF5915',
  },
}

export default theme
