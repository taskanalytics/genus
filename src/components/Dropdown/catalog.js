import Dropdown from '.'
import Button from '../Button'
import {Â Box, Flex } from '../Grid'
import Text from '../Text'
import Heading from '../Heading'

export default {
  title: 'Dropdown',
  path: '/components/dropdown',
  imports: { Dropdown, Button, Box, Flex, Heading, Text },
  component: require('./README.md'),
}
