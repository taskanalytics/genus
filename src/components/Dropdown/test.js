import React from 'react'
import {ThemeProvider} from 'emotion-theming'
import { render, fireEvent, waitForElement, cleanup } from 'react-testing-library'

import theme from '../../styles/theme'
import Dropdown from './index.js'

const renderComponent = (props) => render(
  <ThemeProvider theme={theme}>
    <Dropdown {...props} />
  </ThemeProvider>
)

const actions = [
  {name: 'Act', action: jest.fn()},
  {name: 'Delete', action: jest.fn()},
]

afterEach(cleanup)

describe('<Dropdown> component', () => {
  it('renders trigger button and hidden content', () => {
    const { container } = renderComponent({ actions })
    const node = container.firstChild
    const trigger = node.firstChild
    const dropdown = node.lastChild
    expect(node).toMatchSnapshot()
    expect(node.tagName).toBe('DIV')
    expect(trigger.tagName).toBe('BUTTON')
    expect(trigger.textContent).toBe('Toggle')
    expect(dropdown).toHaveStyle('display: none;')
  })

  it('opens on click', () => {
    const { container } = renderComponent({ actions })
    const trigger = container.querySelector('button:first-child')
    const dropdown = container.firstChild.lastChild

    fireEvent.click(trigger)
    expect(dropdown).toHaveStyle('display: block;')

    const buttons = dropdown.querySelectorAll('button')
    expect(buttons).toHaveLength(2)
    expect(buttons[0].textContent).toBe('Act')
    expect(buttons[1].textContent).toBe('Delete')
  })

  it('calls action fn on action click', () => {
    const data = {
      name: 'Act',
      action: jest.fn(() => true),
    }
    const { container, getByText } = renderComponent({
      actions: [data],
    })

    fireEvent.click(getByText('Act'))

    expect(data.action.mock.calls.length).toBe(1)
    expect(container.firstChild.lastChild).toHaveStyle('display: none;')
  })

  it('closes on click outside', () => {
    const { container, getByText } = renderComponent({ actions })

    // OPen dropdown
    fireEvent.click(getByText('Toggle'))

    expect(container.firstChild.lastChild).toHaveStyle('display: block;')

    // Close with click outside (on container)
    fireEvent.click(container)

    expect(container.firstChild.lastChild).toHaveStyle('display: none;')
  })
})
