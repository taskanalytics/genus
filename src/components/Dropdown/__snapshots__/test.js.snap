// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown> component opens on click 1`] = `
<div>
  <div
    class="css-1r3m7kt"
    data-genus="Dropdown"
  >
    <button
      open=""
      type="button"
    >
      Toggle
    </button>
    <div
      class="css-inpltq"
      data-genus="Box"
      open=""
    >
      <button
        class="css-ms7u4k"
        data-genus="Button"
        type="button"
      >
        X
      </button>
      <button
        class="css-nmwv8r"
      >
        Act
      </button>
      <button
        class="css-nmwv8r"
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;

exports[`<Dropdown> component renders trigger button and no content 1`] = `
<div
  class="css-1r3m7kt"
  data-genus="Dropdown"
>
  <button
    type="button"
  >
    Toggle
  </button>
</div>
`;
